fog/report/validation/src/lib.rs

```diff
@@ -57,3 +68,92 @@ pub struct FullyValidatedFogPubkey {
     /// transaction larger than this limit if the fog pubkey is used.
     pub pubkey_expiry: u64,
 }
+
+/// Represents a set of fully validated fog public keys
+///
+/// These can be fetched before building a transaction, then the rest of the
+/// transaction building process can happen offline.
+#[derive(Default, Debug)]
+pub struct FullyValidatedFogPubkeys {
+    /// Map from fog_url, fog_report_id strings to fully validated pubkeys
+    keys: BTreeMap<(String, String), FullyValidatedFogPubkey>,
+}
+
+impl FullyValidatedFogPubkeys {
+    /// Check if we have a fog pubkey needed for a given recipient.
+    /// If so, do nothing.
+    /// If not, try to fetch one using the resolver.
+    pub fn add_key_for_recipient<R: FogPubkeyResolver>(
+        &mut self,
+        resolver: &R,
+        recipient: &PublicAddress,
+    ) -> Result<(), R::Error> {
+        if let Some(table_key) = self.table_key(recipient) {
+            match self.keys.entry(table_key) {
+                Entry::Occupied(_) => {}
+                Entry::Vacant(ent) => {
+                    ent.insert(resolver.get_fog_pubkey(recipient)?);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Alternative to add_key_for_recipient, this may be helpful for SDK code which doesn't want to make blocking calls into rust.
+    pub fn add_key(&mut self, report_url: String, validated_key: FullyValidatedFogPubkey) {
+        self.keys.insert(
+            (report_url, validated_key.fog_report_id.clone()),
+            validated_key,
+        );
+    }
+
+    /// Construct the key to use in the table for given public address
+    /// Returns None if the public address doesn't have a fog report url.
+    fn table_key(&self, addr: &PublicAddress) -> Option<(String, String)> {
+        addr.fog_report_url().map(|url| {
+            (
+                url.to_string(),
+                addr.fog_report_id().unwrap_or("").to_string(),
+            )
+        })
+    }
+
+    /// Get the smallest pubkey_expiry value of any of these fog pubkeys.
```

Comment:
Why do you want the smallest expiry?  Doesn't it make more sense to use the key that expires later?

Reply:
Oh, I think I see what's going on.  You aren't selecting between keys for an individual user, you're trying to determine an appropriate tombstone block based on all the keys needed for an individual transaction.